### PR TITLE
PHPStan 0.11 and PluginManagers refactor

### PR DESCRIPTION
**Pull request by Slamdunk**
Ref: https://github.com/bitwombat/phpstan-zend-framework/issues/3

- [x] Support PHPStan 0.11
- [x] Support custom ServiceManager injection
- [x] Support main ServiceManager services
- [x] Support main ControllerManager services
- [x] Support main ControllerPluginManager services
- [x] Support main ViewHelperManager services
- [x] Support main RoutePluginManager services

See [ModuleManagerFactory](https://github.com/zendframework/zend-mvc/blob/master/src/Service/ModuleManagerFactory.php)

### CODE REVIEW

File: src/Type/Zend/ServiceManagerGetDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+        $argType = $scope->getType($methodCall->args[0]->value);
+        if (! $argType instanceof ConstantStringType) {
+            return new MixedType();
+        }
+
+        $objectName     = $argType->getValue();
+        $serviceManager = $this->serviceManagerLoader->getServiceManager();
+        if (! $serviceManager->has($objectName)) {
+            return new NullType();
```
Comment (Slamdunk) [inline src/Type/Zend/ServiceManagerGetDynamicReturnTypeExtension.php]: @ondrejmirtes this resembles [Doctrine/GetRepositoryDynamicReturnTypeExtension:58](https://github.com/phpstan/phpstan-doctrine/blob/master/src/Type/Doctrine/GetRepositoryDynamicReturnTypeExtension.php#L58): if I don't check the existence of the service and I directly try to get it, like phpunit-doctrine does, I get a not so user friendly error like:

```
 ------ --------------------------------------------------------------------------------------------------------------------- 
  Line   serviceManagerDynamicReturn.php                                                                                      
 ------ --------------------------------------------------------------------------------------------------------------------- 
         Internal error: Unable to resolve service "bar" to a factory; are you certain you provided it during configuration?  
         Run PHPStan with --debug option and post the stack trace to:                                                         
         https://github.com/phpstan/phpstan/issues/new                                                                        
 ------ --------------------------------------------------------------------------------------------------------------------- 
```
For the code https://github.com/Slamdunk/phpstan-zend-framework/pull/1/files#diff-850327538197c025d0724fadcc005fbbR26

I would like to customize the error message, and most important, get the line that generate the error.
Is this feasible? The final result I would like to achieve is:
```
 ------ ------------------------------------------------------------------------------------------------------
  Line   serviceManagerDynamicReturn.php                                                                      
 ------ ------------------------------------------------------------------------------------------------------
    26   Unable to resolve service "bar" to a factory; are you certain you provided it during configuration?  
 ------ ------------------------------------------------------------------------------------------------------
```

Comment (ondrejmirtes) [reply src/Type/Zend/ServiceManagerGetDynamicReturnTypeExtension.php]: You need not to throw an exception from the extension, and you need to write a rule that checks for this situation.

Comment (Slamdunk) [reply src/Type/Zend/ServiceManagerGetDynamicReturnTypeExtension.php]: Thank you :+1:

File: src/Rules/Zend/ServiceManagerGetMethodCallRule.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     * @var ServiceManagerLoader
+     */
+    private $serviceManagerLoader;
+
+    /**
+     * @var array
+     */
+    private $knownUnmappedAliasToClassServices = [
+        ControllerManager::class   => 'ControllerManager',
+        HelperPluginManager::class => 'ViewHelperManager',
```
Comment (Slamdunk) [inline src/Rules/Zend/ServiceManagerGetMethodCallRule.php]: @weierophinney I'm writing this extension to https://github.com/phpstan/phpstan that checks the correct usage of configured services and plugins.

Everything works flawlessly, expect for services not mapped with their class to the common name, like this two.
So for example if a user writes this:
```php
/** @var \Zend\View\HelperPluginManager $viewHelperManager */
$viewHelperManager = $whateverGetter();
```
Without these two lines I'm commenting at, I wouldn't be able to recognize the service.

Looking at the source code of each Zend component, it appears that there isn't any specific rule you've redacted to enforce `xxxPlugin::class => 'XxxPluginManager'` map.

Another example, `ControllerPluginManager` is mapped:

https://github.com/zendframework/zend-mvc/blob/9f092ff24945f208047d99dff81f67de8862b5f4/src/Service/ServiceListenerFactory.php#L54

`ControllerManager` isn't mapped:

https://github.com/zendframework/zend-mvc/blob/master/src/Service/ServiceListenerFactory.php#L66

How should I handle this topic?
Do I have to assume that there will be no rule, and thus should I maintain this custom list, or maybe it is a bug to be reported in the related component?

**Pull request merged**